akka-actor/src/main/java/akka/dispatch/AbstractNodeQueue.java

```diff
@@ -47,12 +47,12 @@ protected AbstractNodeQueue() {
     protected final Node<T> peekNode() {
         final Node<T> tail = ((Node<T>)Unsafe.instance.getObjectVolatile(this, tailOffset));
         Node<T> next = tail.next();
-        if (next == null && get() != tail) {
+        while (next == null && get() != tail && next != tail.next()) {
```

Comment:
Anyway, at may case everything was bad because `next` was `tail.next()` and it looping forewer.

Reply:
and access to atomic value is cheap. Is very cheap. It is just a function that's return a volatile field from object. It is a bit more complicated, but performance is close to this.

Reply:
This hot spin is carefully designed so that it doesn't do any non-necessary work, trigger cache flush etc. since it is the default mailbox implementation of Akka actors and is also used in a few other places where performance is important and we know there is a single consumer. 

Claiming that you can touch a volatile in each spin (or a yield for that part) would have to be backed up by carefully benchmarking the change in multiple scenarios.

Reply:
@johanandren I'm ready to bet that this file isn't carefully designed. Or at least it has one contradiction.

For example at beginning of this file you can see a comment:
```
 * This queue could be wait-free (i.e. without the spinning loops in peekNode and pollNode) if
 * it were permitted to return null while the queue is not quite empty anymore but the enqueued
 * element is not yet visible. This would break actor scheduling, though.
```

How you can see original implementation of `peekNode()` and `pollNode()` isn't wait-free.

Reply:
It's not wait free, it is lock free.

Reply:
@johanandren I definitely agree that it is lock-free.

Anyway, the comment exists at this code since https://github.com/akka/akka/commit/f9d3789bfca6eb08fb2dcf9cc7b1dc1643e099e6